### PR TITLE
Simplify L1 interaction

### PR DESCRIPTION
**Pull request by sergey-melnychuk**
Closes https://github.com/eqlabs/pathfinder/issues/1009

### CODE REVIEW

**Comment by sergey-melnychuk** [comment]
@Mirko-von-Leipzig @CHr15F0x @kkovaacs PTAL

File: crates/ethereum/src/lib.rs [hunk 1]
```diff
@@ -1,111 +1,295 @@
...
+    let mut output: [u8; 32] = Default::default();
+    keccak_hash::keccak_256(signature, &mut output[..]);
+    format!("0x{}", hex::encode(&output[0..4]))
+}
+
+fn get_h256(value: &serde_json::Value) -> anyhow::Result<H256> {
+    use std::str::FromStr;
+    value
+        .as_str()
+        .and_then(|txt| H256::from_str(txt).ok())
```
Comment (CHr15F0x) [inline crates/ethereum/src/lib.rs:157]: is non-fixed hex str length an issue here? ie. can the abi return a non 32 hex digit string and then does H256 support it?

Comment (sergey-melnychuk) [reply crates/ethereum/src/lib.rs:157]: Why would it be?

https://docs.rs/primitive-types/latest/primitive_types/struct.H256.html#impl-FromStr-for-H256

https://docs.rs/fixed-hash/latest/src/fixed_hash/hash.rs.html#576

File: crates/storage/src/schema/revision_0034.rs
Comment (Mirko-von-Leipzig) [inline crates/storage/src/schema/revision_0034.rs:1]: I think we should populate this new table with some entries. We can take the existing `L1 == L2` points from the refs table and fill in the details here, so that on startup we already have up-to-date information from the last time we started.

File: crates/common/src/lib.rs [hunk 2]
```diff
@@ -374,9 +374,9 @@ impl std::ops::SubAssign<u64> for BlockNumber {
     }
 }
 
-impl From<EthereumBlockNumber> for ethers::types::BlockId {
+impl From<EthereumBlockNumber> for BlockId {
```
Comment (Mirko-von-Leipzig) [inline crates/common/src/lib.rs:377]: It's likely we can delete the `EthereumXXX` types now? Let me check the rest of the PR..

File: crates/ethereum/src/lib.rs (hunk 1 above)
Comment (sergey-melnychuk) [reply crates/ethereum/src/lib.rs]: Nice catch, thanks.

Done: https://github.com/eqlabs/pathfinder/pull/1110/commits/f6032a68e82d0ca1c1a8d35fbe71c360bfecf429

Comment (Mirko-von-Leipzig) [reply crates/ethereum/src/lib.rs]: Non-spec compliant eth servers can do whatever they wish. In which case this will error I think? Which is fine -- don't think we need to cater for short (or long) strings.

File: crates/ethereum/src/lib.rs [hunk 3]
```diff
@@ -1,111 +1,354 @@
...
+    keccak_hash::keccak_256(signature, &mut output[..]);
+    format!("0x{}", hex::encode(&output[0..4]))
+}
+
+fn get_h256(value: &serde_json::Value) -> anyhow::Result<H256> {
+    use std::str::FromStr;
+    value
+        .as_str()
+        .map(lpad64)
+        .and_then(|val| H256::from_str(&val).ok())
```
Comment (Mirko-von-Leipzig) [inline crates/ethereum/src/lib.rs:158]:
```suggestion
        .ok_or(anyhow::anyhow!("Failed to parse H256"))
```

File: crates/ethereum/src/lib.rs [hunk 4]
```diff
@@ -1,111 +1,354 @@
...
+        .ok_or(anyhow::anyhow!("Failed to fetch H256"))
+}
+
+fn get_u256(value: &serde_json::Value) -> anyhow::Result<U256> {
+    use std::str::FromStr;
+    value
+        .as_str()
```
Comment (Mirko-von-Leipzig) [inline crates/ethereum/src/lib.rs:165]:
```suggestion
        .ok_or(anyhow::anyhow!("Failed to parse U256"))
```

File: crates/ethereum/src/lib.rs [hunk 5]
```diff
@@ -1,111 +1,354 @@
...
+}
 
-    fn try_from(log: &ethers::types::Log) -> Result<Self, Self::Error> {
-        let block = BlockOrigin::try_from(log)?;
-        let transaction = TransactionOrigin::try_from(log)?;
-        let log_index = log.log_index.context("missing log index")?.as_u64();
+fn get_felt(value: H256) -> anyhow::Result<Felt> {
+    let felt = Felt::from_be_slice(value.as_bytes())?;
+    Ok(felt)
+}
```
Comment (Mirko-von-Leipzig) [inline crates/ethereum/src/lib.rs:175]:
```suggestion
    BlockNumber::new(value).ok_or(anyhow::anyhow!("Overflow while converting U256 to u64"))
```

File: crates/pathfinder/src/bin/pathfinder/main.rs [hunk 6]
```diff
@@ -368,28 +365,28 @@ mod pathfinder_context {
                     network_id: ChainId::MAINNET,
                     gateway: GatewayClient::mainnet(),
                     database: data_directory.join("mainnet.sqlite"),
-                    l1_core_address: Self::MAINNET_CORE,
+                    l1_core_address: EthereumAddress(H160::from(core_addr::MAINNET)),
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/bin/pathfinder/main.rs:368]: Maybe its worth moving these definitions to `EthereumAddress`? e.g. `EthereumAddress::MAINNET` etc. This would require moving them `common` though as otherwise I expect there will be a type cycle or two.. Just a thought, not a strong opinion either way.

File: crates/pathfinder/src/state/sync.rs [hunk 7]
```diff
@@ -471,72 +411,27 @@ async fn update_sync_status_latest(
...
-        let l1_l2_head = RefsTable::get_l1_l2_head(&transaction).context("Query L1-L2 head")?;
-        match l1_l2_head {
-            Some(head) if head >= reorg_tail => {
-                let new_head = match reorg_tail {
-                    BlockNumber::GENESIS => None,
-                    other => Some(other - 1),
-                };
-                RefsTable::set_l1_l2_head(&transaction, new_head).context("Update L1-L2 head")?;
+        if let Some(l2_root) = l2_root {
+            if l2_root == update.global_root {
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/state/sync.rs:431]: This keeps the existing semantics of using the state commitment as the point of comparison. We should switch this to using the block hash instead as it is a stronger guarantee (since it includes the state commitment). But we can do this in a separate PR if you want.

Comment (Mirko-von-Leipzig) [reply crates/pathfinder/src/state/sync.rs:431]: We should also definitely log if `L1 != L2` probably as `ERROR` or at the very least `WARN`. 

In addition, we might want to consider attempting a repair i.e. finding the last point at which `L1 == L2` still holds true, by searching backwards through the `L1 state table`.

File: crates/pathfinder/src/state/sync/l1.rs [hunk 8]
```diff
@@ -1,680 +1,49 @@
...
-    transport: T,
+    tx_event: mpsc::Sender<EthereumStateUpdate>,
+    ethereum: T,
     chain: Chain,
     core_address: H160,
-    head: Option<StateUpdateLog>,
 ) -> anyhow::Result<()>
 where
-    T: EthereumTransport + Send + Sync + Clone,
+    T: EthereumApi + Send + Sync + Clone,
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/state/sync/l1.rs:18]: Bounds are more restrictive than `sync_impl`

File: crates/pathfinder/src/state/sync/l1.rs [hunk 9]
```diff
@@ -1,680 +1,49 @@
...
-    }
-
-    /// Sends [Event::Reorg] on its channel.
-    async fn reorg(&self, block: BlockNumber) -> Result<(), ChannelClosedError> {
-        self.0
-            .send(Event::Reorg(block))
-            .await
-            .map_err(|_send_err| ChannelClosedError)
-    }
+    sync_impl(ethereum, tx_event, chain, core_address).await
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/state/sync/l1.rs:21]: Do we need the extra indirection / abstraction?

**Review by Mirko-von-Leipzig — COMMENTED** [review]
Looks good to me - just a few comments I'd like to discuss before approving. Nothing major standing in the way though!

File: crates/ethereum/src/lib.rs (hunk 3 above)
Comment (sergey-melnychuk) [reply crates/ethereum/src/lib.rs:158]: already there, please check next commits

File: crates/ethereum/src/lib.rs (hunk 4 above)
Comment (sergey-melnychuk) [reply crates/ethereum/src/lib.rs:165]: already there, please check next commits

File: crates/ethereum/src/lib.rs (hunk 5 above)
Comment (sergey-melnychuk) [reply crates/ethereum/src/lib.rs:175]: already there, please check next commits

File: crates/pathfinder/src/bin/pathfinder/main.rs (hunk 6 above)
Comment (sergey-melnychuk) [reply crates/pathfinder/src/bin/pathfinder/main.rs:368]: They won't be `const` anyway. But I think this is a good place to get rid of `EthereumAddress` wrapper.

File: crates/ethereum/src/lib.rs (hunk 3 above)
Comment (Mirko-von-Leipzig) [reply crates/ethereum/src/lib.rs:158]: Ah sorry, think I grabbed the wrong line. I just meant the phrasing is slightly wrong for this function since its parsing and not fetching. Same for the other.

File: crates/pathfinder/src/state/sync/l1.rs (hunk 8 above)
Comment (sergey-melnychuk) [reply crates/pathfinder/src/state/sync/l1.rs:18]: I didn't change the bounds, only the trait name.

Comment (sergey-melnychuk) [reply crates/pathfinder/src/state/sync/l1.rs]: Relaxed the restricted version to match `sync`.

File: crates/pathfinder/src/state/sync/l1.rs (hunk 9 above)
Comment (sergey-melnychuk) [reply crates/pathfinder/src/state/sync/l1.rs]: Done

File: crates/storage/src/schema/revision_0034.rs
Comment (sergey-melnychuk) [reply crates/storage/src/schema/revision_0034.rs:1]: Updated migration, if the refs is pointing to the recent head, if will be copied into l1_state.

File: crates/storage/src/schema/revision_0034.rs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            .map(|(storage, class)| StateCommitment::calculate(storage, class));
+            hash.zip(state).map(|(hash, state)| EthereumStateUpdate {
+                global_root: state,
+                block_number: number,
+                block_hash: hash,
+            })
+        } else {
+            None
+        })
+    })
```
Comment (Mirko-von-Leipzig) [inline crates/storage/src/schema/revision_0034.rs:49]: If possible, I think doing this manually here is good i.e. copy the sql over instead. What ends up happening is that changes in the storage crate end up impacting the migrations -- and sometimes silently without noticing.

File: crates/pathfinder/src/state/sync.rs (hunk 7 above)
Comment (sergey-melnychuk) [reply crates/pathfinder/src/state/sync.rs]: Added a hash match check and a `WARN`. If hashes do not match (even though "it should never happen" (c)), then the most recent block where hashes did match must be in refs table, as l1_state and refs are updated simultaneously.

File: crates/storage/src/schema/revision_0034.rs (hunk 10 above)
Comment (sergey-melnychuk) [reply crates/storage/src/schema/revision_0034.rs:46]: I am about to write some tricky join...

File: crates/storage/src/schema/revision_0034.rs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    StarknetBlocksTable,
+};
+
+pub(crate) fn migrate(tx: &Transaction<'_>) -> anyhow::Result<()> {
+    tx.execute("DROP TABLE l1_state", [])?;
+    tx.execute(
+        r"CREATE TABLE l1_state (
+            starknet_block_number      INTEGER PRIMARY KEY,
+            starknet_block_hash        BLOB    NOT NULL,
+            starknet_global_root       BLOB    NOT NULL
```
Comment (kkovaacs) [inline crates/storage/src/schema/revision_0034.rs:16]: Hmm -- shouldn't we use `starknet_state_commitent` instead of `starknet_global_root`? At least that seems to be the official name now: https://docs.starknet.io/documentation/architecture_and_concepts/State/starknet-state/#state_commitment

File: crates/ethereum/src/lib.rs [hunk 12]
```diff
@@ -1,111 +1,354 @@
...
+        Decoder::Hex.decode(b"d5c325D183C592C94998000C5e0EED9e6655c020");
+}
 
-#[derive(Debug, Clone, PartialEq, Hash, Eq)]
-pub struct BlockOrigin {
-    pub hash: EthereumBlockHash,
-    pub number: EthereumBlockNumber,
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EthereumStateUpdate {
+    pub global_root: StateCommitment,
```
Comment (kkovaacs) [inline crates/ethereum/src/lib.rs:17]: Maybe `state_commitment` instead of `global_root` here? (https://docs.starknet.io/documentation/architecture_and_concepts/State/starknet-state/#state_commitment)

File: crates/storage/src/schema/revision_0034.rs (hunk 10 above)
Comment (sergey-melnychuk) [reply crates/storage/src/schema/revision_0034.rs]: Not that tricky join after all.

File: crates/ethereum/src/lib.rs (hunk 12 above)
Comment (sergey-melnychuk) [reply crates/ethereum/src/lib.rs]: The [contract's](https://etherscan.io/address/0xc662c410C0ECf747543f5bA90660f6ABeBD9C8c4#readProxyContract) method is `stateRoot()`, I suggest we keep naming consistent. It is clear though what `StateCommitment` stands for.

Comment (Mirko-von-Leipzig) [reply crates/ethereum/src/lib.rs]: The reason the L1 calls it `stateRoot` is because its what they called it originally. I imagine if they could redo it now, they would call it `stateCommitment`

**Review by Mirko-von-Leipzig — APPROVED** [review]

**Review by kkovaacs — APPROVED** [review]
